Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -220,73 +220,81 @@ struct Gx2FitterResult {
-               Gx2FitterResult<traj_t>& result, const Logger& logger) {
-  auto predicted = trackStateProxy.predicted();
-  auto measurement = trackStateProxy.template calibrated<measDim>();
+/// @param aMatrix The aMatrix sums over the second derivatives
+/// @param bVector The bVector sums over the first derivatives
+/// @param chi2sum The total chi2 of the system
+/// @param jacobianFromStart The Jacobian matrix from start to the current state
+/// @param trackState The track state to analyse
+/// @param logger A logger instance
+template <std::size_t kMeasDim, typename track_state_t>
+void addToGx2fSums(BoundMatrix& aMatrix, BoundVector& bVector, double& chi2sum,
+                   const BoundMatrix& jacobianFromStart,
+                   const track_state_t& trackState, const Logger& logger) {
+  auto predicted = trackState.predicted();
+  auto measurement = trackState.template calibrated<kMeasDim>();
   auto covarianceMeasurement =
-      trackStateProxy.template calibratedCovariance<measDim>();
-  // Project Jacobian and predicted measurements into the measurement dimensions
-  auto projJacobian = (trackStateProxy.projector()
-                           .template topLeftCorner<measDim, eBoundSize>() *
-                       result.jacobianFromStart)
-                          .eval();
-  auto projPredicted = (trackStateProxy.projector()
-                            .template topLeftCorner<measDim, eBoundSize>() *
-                        predicted)
-                           .eval();
-
-  ACTS_VERBOSE("Processing and collecting measurements in Actor:"
-               << "\n    Measurement:\t" << measurement.transpose()
-               << "\n    Predicted:\t" << predicted.transpose()
-               << "\n    Projector:\t" << trackStateProxy.effectiveProjector()
-               << "\n    Projected Jacobian:\t" << projJacobian
-               << "\n    Covariance Measurements:\t" << covarianceMeasurement);
-
-  // Collect residuals, covariances, and projected jacobians
-  for (std::size_t i = 0; i < measDim; i++) {
-    if (covarianceMeasurement(i, i) < 1e-10) {
-      ACTS_WARNING("Invalid covariance of measurement: cov(" << i << "," << i
-                                                             << ") ~ 0")
-      continue;
-    }
-
-    result.collectorResiduals.push_back(measurement[i] - projPredicted[i]);
-    result.collectorCovariances.push_back(covarianceMeasurement(i, i));
-    result.collectorProjectedJacobians.push_back(projJacobian.row(i));
-
-    ACTS_VERBOSE("    Splitting the measurement:"
-                 << "\n        Residual:\t" << measurement[i] - projPredicted[i]
-                 << "\n        Covariance:\t" << covarianceMeasurement(i, i)
-                 << "\n        Projected Jacobian:\t" << projJacobian.row(i));
+      trackState.template calibratedCovariance<kMeasDim>();
+  auto projector =
+      (trackState.projector().template topLeftCorner<kMeasDim, eBoundSize>())
+          .eval();
```

Comment:
In general I would be careful with using `auto` in the context of Eigen https://eigen.tuxfamily.org/dox-3.3/TopicPitfalls.html

Usually we try to type Eigen object explicitly. What you could to here is
```suggestion
  ActsMatrix<kMeasDim, eBoundSize> projector = trackState.projector().template topLeftCorner<kMeasDim, eBoundSize>();
```

similar below

Reply:
Yes, I know. Here it still worked quite fine, but I had during testing already cases where some compilers failed. Let's make it more stable.